GVFS/GVFS.Common/GitHubUpgrader.cs

```diff
@@ -529,29 +529,65 @@ public bool TryLoad(out string error)
+                // upgrade.ring and upgrade.feedURL. upgrade.feedURL will override upgrade.ring
+                // config. If upgrade.feedURL is set, but does not contain valid ring info then
+                // upgrade ring would be set to NoConfig and auto upgrade checks will fail.
+                if (!this.LocalConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.UpgradeFeedUrl, out upgradeFeedUrl, out error))
                 {
-                    RingType ringType;
-                    if (Enum.TryParse(ringConfig, ignoreCase: true, result: out ringType) &&
-                        Enum.IsDefined(typeof(RingType), ringType) &&
-                        ringType != RingType.Invalid)
-                    {
-                        this.UpgradeRing = ringType;
-                    }
-                    else
+                    error = loadError;
+                    return false;
+                }
+
+                // If upgradeFeedUrl is available then use it for the ring config, if not use regular
+                // upgrade.ring.
+                if (!string.IsNullOrEmpty(upgradeFeedUrl))
+                {
+                    this.UpgradeRing = RingType.None;
+
+                    string[] expectedRings = new string[] { RingType.Slow.ToString(), RingType.Fast.ToString() };
+                    foreach (string ring in expectedRings)
                     {
-                        if (!string.IsNullOrEmpty(ringConfig))
+                        string ringSubstring = $"@{ring}/";
+                        if (upgradeFeedUrl.IndexOf(ringSubstring, StringComparison.OrdinalIgnoreCase) != -1)
                         {
-                            this.UpgradeRing = RingType.Invalid;
+                            ringConfig = ring;
+                            break;
                         }
                     }
+                }
+                else if (!this.LocalConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.UpgradeRing, out ringConfig, out error))
+                {
+                    error = loadError;
+                    return false;
+                }
 
-                    return true;
+                this.ParseUpgradeRing(ringConfig);
+                return true;
+            }
+
+            public void ParseUpgradeRing(string ringConfig)
+            {
+                RingType ringType;
+                if (Enum.TryParse(ringConfig, ignoreCase: true, result: out ringType) &&
+                    Enum.IsDefined(typeof(RingType), ringType) &&
+                    ringType != RingType.Invalid)
+                {
+                    this.UpgradeRing = ringType;
+                    return;
                 }
 
-                error = "Could not read GVFS Config." + Environment.NewLine;
-                error += GVFSConstants.UpgradeVerbMessages.SetUpgradeRingCommand;
-                return false;
+                if (string.IsNullOrEmpty(ringConfig))

```

Comment:
Should we check `string.IsNullOrEmpty(ringConfig)` before calling `Enum.TryParse(ringConfig ...`?